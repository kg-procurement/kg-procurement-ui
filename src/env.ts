export const API_BASE_URL =
  import.meta.vitest || process.env.NODE_ENV === 'test'
    ? 'https://www.kg-procurement-testing.com'
    : import.meta.env.VITE_API_BASE_URL
