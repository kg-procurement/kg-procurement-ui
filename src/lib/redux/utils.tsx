import { SerializedError } from '@reduxjs/toolkit'
import { FetchBaseQueryError } from '@reduxjs/toolkit/query'

export type ZodParseError = Extract<
  FetchBaseQueryError,
  {
    status: 'PARSING_ERROR'
    originalStatus: number
    data: string
    error: string
  }
>

export function isZodParseError(
  err: SerializedError | FetchBaseQueryError | undefined,
): err is ZodParseError {
  if (!err) return false
  return (
    'data' in err &&
    'error' in err &&
    'originalStatus' in err &&
    'status' in err
  )
}

export function extractZodError(err: ZodParseError): React.ReactNode {
  const errorList = err.error.split('\n').slice(0, -1)
  return (
    <div className="flex flex-col">
      <div className="font-semibold">
        Response does not match validator schema:
      </div>
      <ul className="list-inside list-disc">
        {errorList.map((e, i) => (
          <li key={i}>{e}</li>
        ))}
      </ul>
    </div>
  )
}

export function isErrorWithMessage(
  err: FetchBaseQueryError,
): err is { status: number, data: { error: string } } {
  if (!err.data) return false
  return (
    typeof err.data === 'object' &&
    'error' in err.data &&
    typeof err.data.error === 'string'
  )
}
