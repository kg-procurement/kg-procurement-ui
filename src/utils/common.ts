export const noop = () => {}

export const formatPrice = (
  amount: number,
  quantityPrice: number,
  currencyCode: string,
  uomName: string
): string => {
  const formattedPrice = new Intl.NumberFormat(
    currencyCode === 'IDR' ? 'id-ID' : 'en-US',
    {
      style: 'currency',
      currency: currencyCode,
    }
  ).format(amount);

  return quantityPrice > 1
    ? `${formattedPrice} / ${quantityPrice} ${uomName}`
    : `${formattedPrice} / ${uomName}`;
};
