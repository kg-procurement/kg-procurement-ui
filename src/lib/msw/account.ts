import { http, HttpResponse } from 'msw'

import { API_BASE_URL } from '@/env.ts'
import { RegisterAccountResponse } from '@/lib/redux/features/account/validation.ts'

export const accountHandlers = [
  http.post(`${API_BASE_URL}/account/register`, () => {
    return HttpResponse.json({
      message: 'Account registered successfully',
    } satisfies RegisterAccountResponse)
  }),
]
