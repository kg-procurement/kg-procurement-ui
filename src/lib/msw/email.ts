import { http, HttpResponse } from 'msw'

import { API_BASE_URL } from '@/env.ts'
import { GetEmailStatusesResponse } from '@/lib/redux/features/email/validator.ts'

export const emailHandlers = [
  http.get(`${API_BASE_URL}/email-status`, () =>
    HttpResponse.json({
      email_status: [
        {
          id: '1',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '10',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '11',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '2',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '3',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '4',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '5',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '6',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '7',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
        {
          id: '8',
          email_to: 'andrewjeremy12345@gmail.com',
          status: 'SENT',
          modified_date: '2024-11-14T09:14:46.914Z',
        },
      ],
      metadata: {
        total_page: 2,
        current_page: 1,
        total_entries: 11,
      },
    } satisfies GetEmailStatusesResponse),
  ),
]
