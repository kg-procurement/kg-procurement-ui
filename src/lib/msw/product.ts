import { http, HttpResponse } from 'msw'

import { API_BASE_URL } from '@/env.ts'
import { UpdateProductResponse } from '@/lib/redux/features/product/validation.ts'

export const productHandlers = [
  http.put(`${API_BASE_URL}/product/:id`, ({ params }) => {
    const { id } = params
    return HttpResponse.json({
      id: Array.isArray(id) ? id[0] : id,
      product_category_id: '38',
      uom_id: '26',
      income_tax_id: '0',
      product_type_id: '3',
      name: 'Juicer',
      description: 'Best Juicer',
      modified_date: '2024-10-17T14:22:28.826745Z',
      modified_by: '0',
    } satisfies UpdateProductResponse)
  }),
]
