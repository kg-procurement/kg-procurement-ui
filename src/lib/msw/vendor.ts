import { http, HttpResponse } from 'msw'

import { API_BASE_URL } from '@/env.ts'
import { GetVendorByIdResponse, GetVendorsResponse, UpdateVendorResponse } from '@/lib/redux/features/vendor/validation.ts'

export const vendorHandlers = [
  http.get(`${API_BASE_URL}/vendor`, () => {
    return HttpResponse.json({
      vendors: [
        {
          id: '2502',
          name: 'Adriza Nikmah, PT',
          description: 'Headset dll',
          bp_id: '2502',
          bp_name: 'Adriza Nikmah, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2507',
          name: 'Toko Amazon',
          description: 'Menjual segala macam Kebutuhan Dapur/ Home App...',
          bp_id: '2507',
          bp_name: 'Amazon',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2022-11-24T10:12:25Z',
          modified_by: '165',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2508',
          name: 'Aneka Warna',
          description: 'Vendor Sablon dan Garment',
          bp_id: '2508',
          bp_name: 'Aneka Warna',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2509',
          name: 'Angkasa Buana Niaga Cipta, PT',
          description: 'Headset Plantronics',
          bp_id: '2509',
          bp_name: 'Angkasa Buana Niaga Cipta, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2511',
          name: 'Apartement Permata Senayan',
          description: '0',
          bp_id: '2511',
          bp_name: 'Apartement Permata Senayan',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2512',
          name: 'Arta Muara',
          description: 'Mesin Barcode',
          bp_id: '2512',
          bp_name: 'Arta Muara',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2513',
          name: 'Asaba Computer Centre',
          description: 'Miscrosoft Gold Partner',
          bp_id: '2513',
          bp_name: 'Asaba Computer Centre',
          rating: 1,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: '102722',
          modified_date: '2022-12-05T11:47:24Z',
          modified_by: '7',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2514',
          name: 'Astra Graphia TBK, PT',
          description: 'Distributor Xerox',
          bp_id: '2514',
          bp_name: 'Astra Graphia TBK, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2515',
          name: 'Asus Service Center',
          description: 'Service Center dan Sales Spare Spart Asus',
          bp_id: '2515',
          bp_name: 'Asus Service Center',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2516',
          name: 'Atrindo Asia Global, PT',
          description: 'Penyemprotan Disinfectant',
          bp_id: '2516',
          bp_name: 'Atrindo Asia Global, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
      ],
      metadata: {
        total_page: 10,
        current_page: 1,
        total_entries: 100,
      },
    } satisfies GetVendorsResponse)
  }),
  http.get(`${API_BASE_URL}/vendor/:id`, (req) => {
    const { id } = req.params
    if (id === '-1') {
      return HttpResponse.json(null, { status: 200 })
    }

    return HttpResponse.json(
      {
        id: '2502',
        name: 'Adriza Nikmah, PT',
        description: 'Headset dll',
        bp_id: '2502',
        bp_name: 'Adriza Nikmah, PT',
        rating: 0,
        area_group_id: '1',
        area_group_name: 'Indonesia',
        sap_code: 'None',
        modified_date: '2020-10-29T13:39:12Z',
        modified_by: '0',
        dt: '2024-08-28T00:00:00Z',
      } satisfies GetVendorByIdResponse,
      { status: 200 },
    )
  }),
  http.put(`${API_BASE_URL}/vendor/:id`, ({ params }) => {
    const { id } = params
    return HttpResponse.json({
      id: Array.isArray(id) ? id[0] : id,
      name: 'Adriza Nikmah, PT',
      description: 'Headset dll',
      bp_id: '2502',
      bp_name: 'Adriza Nikmah, PT',
      rating: 0,
      area_group_id: '1',
      area_group_name: 'Indonesia',
      sap_code: 'None',
      modified_date: '2020-10-29T13:39:12Z',
      modified_by: '0',
      dt: '2024-08-28T00:00:00Z',
    } satisfies UpdateVendorResponse)
  }),
]
