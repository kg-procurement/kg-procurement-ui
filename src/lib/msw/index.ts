import { setupServer } from 'msw/node'

import { productHandlers } from '@/lib/msw/product.ts'
import { vendorHandlers } from '@/lib/msw/vendor.ts'

const handlers = [...vendorHandlers, ...productHandlers]

export const mswServer = setupServer(...handlers)
