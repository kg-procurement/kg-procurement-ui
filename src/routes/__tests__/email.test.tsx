import { render, screen } from '@testing-library/react'
import { http, HttpResponse } from 'msw'

import { API_BASE_URL } from '@/env.ts'
import { mswServer } from '@/lib/msw/index.ts'
import { withWrappers } from '@/lib/testing/utils.tsx'
import { waitForNoLoadingOverlay } from '@/lib/testing/wait-for.ts'

import EmailPage from '../email.tsx'

describe('<EmailPage />', () => {
  it('should render properly without crashing', () => {
    render(withWrappers(<EmailPage />))
  })

  it('should render the header section with the logo', () => {
    render(withWrappers(<EmailPage />))

    const logo = screen.getByAltText('Kompas Gramedia Logo Background')
    expect(logo).toBeInTheDocument()
    expect(logo).toHaveAttribute('src', '/kompas-gramedia-logo-bg.svg')

    expect(screen.getByText('Email Status')).toBeInTheDocument()
  })

  it('should render email status content properly', async () => {
    render(withWrappers(<EmailPage />, { withRoot: true }))
    await waitForNoLoadingOverlay()
    expect(screen.getByTestId('email-status-table').innerText).toMatchSnapshot()
  })

  it('should render input boxes for filtering by email', () => {
    render(withWrappers(<EmailPage />))

    const emailFiterInput = screen.getByTestId('filter-by-email')

    expect(screen.getByPlaceholderText('Filter by Email')).toBeInTheDocument()
    expect(emailFiterInput).toBeInTheDocument()
  })

  it('should call update email status mutation when clicking the save button', async () => {
    render(withWrappers(<EmailPage />, { withRoot: true }))
    await waitForNoLoadingOverlay()

    const ids = [
      'email-status-button-failed',
      'email-status-button-success',
      'email-status-button-inprogress',
      'email-status-button-completed',
    ]
    ids.forEach((id) => {
      const button = screen.getAllByTestId(id)[0]
      expect(button).toBeInTheDocument()
      button.click()
    })
  })

  it('should call handle when update email status mutation fail', async () => {
    const statusCode = 500
    const errorMessage =
      'error'
    mswServer.use(
      http.put(`${API_BASE_URL}/email-status/:id`, () =>
        HttpResponse.json(
          {
            error: errorMessage,
          },
          { status: statusCode },
        ),
      ),
    )

    render(withWrappers(<EmailPage />, { withRoot: true }))
    await waitForNoLoadingOverlay()

    const ids = [
      'email-status-button-failed',
      'email-status-button-success',
      'email-status-button-inprogress',
      'email-status-button-completed',
    ]
    ids.forEach((id) => {
      const button = screen.getAllByTestId(id)[0]
      expect(button).toBeInTheDocument()
      button.click()
    })
  })
})
