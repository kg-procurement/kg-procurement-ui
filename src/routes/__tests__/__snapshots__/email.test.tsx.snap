// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EmailPage /> > should render email status content properly 1`] = `"Vendor NameEmail ToPerformance ScoreDate SentStatusMulti Kharisma Solusindo, PTAndrewjeremy12345@gmail.com-1002024-12-04T23:19:43.14054ZFailedSuccessIn ProgressCompletedMulti Kharisma Solusindo, PTAndrewjeremy12345@gmail.com-1002024-12-05T00:11:12.155375ZFailedSuccessIn ProgressCompletedFors Fortis Indonesia, PTkiminonamaewa98@gmail.com02024-12-05T00:11:37.379726ZFailedSuccessIn ProgressCompletedMulti Kharisma Solusindo, PTAndrewjeremy12345@gmail.com-1002024-12-05T00:11:37.935501ZFailedSuccessIn ProgressCompleted"`;
